### PR TITLE
Drive devtools survey from metadata hosted on flutter.dev

### PR DESCRIPTION
**Pull request by kenzieschmoll**
Screen Shot 2020-11-02 at 3 23 00 PM

Survey data is pulled from here: https://flutter.dev/f/dart-devtools-survey-metadata.json

### CODE REVIEW

File: packages/devtools_app/lib/src/notifications.dart [hunk 1]
```diff
@@ -235,19 +267,65 @@ class _NotificationState extends State<_Notification>
...
+      maxLines: 6,
+    );
+  }
+
+  Widget _buildActions() {
+    if (widget.actions.isEmpty) return const SizedBox();
+    return Row(
+      mainAxisAlignment: MainAxisAlignment.end,
+      children: [
+        for (int i = 0; i < widget.actions.length; i++) ...[
```
Comment (jacob314) [inline packages/devtools_app/lib/src/notifications.dart:299]: we should create a joinWith(List<Widget>, Widget separator) helper method or maybe extension method on List<Widget> as we've got this pattern a few places.

File: packages/devtools_app/lib/src/survey.dart [hunk 2]
```diff
@@ -0,0 +1,166 @@
+import 'dart:convert';
+
+import 'dart:io' if (dart.library.html) 'dart:html';
```
Comment (jacob314) [inline packages/devtools_app/lib/src/survey.dart:3]: this makes me really nervous as a conditional import. dart:html and dart:io are super different. Can we remove this asnd use package:http and similar?

File: packages/devtools_app/lib/src/survey.dart [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+  /// Duration for which we should show the survey notification.
+  ///
+  /// We use a very long time here to give the appearance of a persistent
+  /// notification. The user will need to interact with the prompt to dismiss
+  /// it.
+  static const notificationDuration = Duration(days: 1);
+
+  /// The amount of time (one day) to wait between checks to the file at
+  /// [metadataUrl].
+  static const checkInterval = Duration(days: 1);
```
Comment (jacob314) [inline packages/devtools_app/lib/src/survey.dart:32]: why do we bother with this? just loading the file and relying on browser caching should be fine as long as we don't block the devtools UI waiting for the json file to be loaded.

File: packages/devtools_app/lib/src/survey.dart [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+    final currentTimeMs = DateTime.now().millisecondsSinceEpoch;
+    final activeSurveyRange = Range(
+      _cachedSurvey.startDate.millisecondsSinceEpoch,
+      _cachedSurvey.endDate.millisecondsSinceEpoch,
+    );
+    return activeSurveyRange.contains(currentTimeMs);
+  }
+
+  Future<bool> shouldFetchSurveyContent() async {
```
Comment (jacob314) [inline packages/devtools_app/lib/src/survey.dart:96]: lets remove this method and just rely on browser caching. loading one tiny json file from a server on devtools startup is completely fine if we don't block devtools startup to do it. if we are concerned we can tweak caching settings on the flutter.dev website to keep it cached for a day.

**Review by jacob314 — COMMENTED** [review]
done with a round of comments.

File: packages/devtools_app/lib/src/survey.dart (hunk 2 above)
Comment (kenzieschmoll) [reply packages/devtools_app/lib/src/survey.dart:3]: The only reason this is here is for use of the `HttpStatus.ok` const, but we could just as easily write `200`.

Comment (jacob314) [reply packages/devtools_app/lib/src/survey.dart:3]: lets write 200 or maybe package:http does or should define that const.
dart static checking falls over badly for conditional imports like this.

**Review by jacob314 — COMMENTED** [review]
The logic for the survey shown count will run into trouble the second time we need to show a survey. Won't we use the survey shown count from the previous server id? Maybe change the survey shown count api to take a survey id so we keep track of the # of times each survey id has been shown.

**Comment by kenzieschmoll** [comment]
These values are already stored by the active survey, so if there is a new survey, a new entry will be created in storage and that entry will have a surveyShownCount 0 and surveyActionTaken false

**Comment by jacob314** [comment]
lgtm

**Pull request merged**